### PR TITLE
TP part-time templates update - part 1

### PR DESCRIPTION
**Pull request by khunshwephyo**
---
name: Certificate Template Addition
about: This is the workflow for requesting new certificate templates to be added to
  the OpenCerts repository
title: "[New Template]"
labels: new template
assignees: ''

---

# Pull Request Guidelines for Adding Certificate Templates
This document is a work in progress but here are some basic checks. As these are only basic guidelines, meeting the below doesn't indicate there will be no issues with your pull request.

### Pre-merge checks

- [x] Did not modify any files outside of your organisation's template folder (e.g package-lock.json or anything else)
- [x] Ensure that your code has been [rebased](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) on top of latest OpenCerts master
- [x] Linter issues resolved (Run `npm run lint:fix` to see issues)
- [x] `npm run test` passes
- [x] `npm run test:integration` passes
- [x] [Travis Build passes](https://docs.travis-ci.com/user/for-beginners/)

### Certificate Template 
- [x] No more than 5 templates or 25 added/modified files in the pull request
- [x] Ensure that your .opencert file's data complies with the intentions of the OpenCerts' schema - e.g recipient related information is inside the `recipient` object, etc.
- [x] Integration test for each template that checks that the correct rendering is done given a sample certificate
- [x] Sample certificate file included for each template, located alongside the integration test for each template
- [x] Sample certificates must obviously be a sample certificate
  - [x] Obviously fictitious name
  - [ ] Obviously sample signatory images
- [x] No fixed-size raster images as part of certificate layout
- [x] Mobile responsive design
- [x] Date parsing should be localised to template author's timezone
- [x] Webpack chunking code is correct
  - [x] Has chunking code
  - [x] Same chunking code as the other certificates belonging to that institute
- [x] Certificate Store Addresses have been updated
  - [x] Template Whitelist
  - [x] Registry
- [x] Template should not be using resources(images etc.) on the website outside of their own folder (e.g institute logo shouldn't be used from /static because there's no guarantee it will not change)

### CODE REVIEW

**Comment by khunshwephyo** [comment]
My apologies for having too many files in this pull request.  I tried my best to reduce number of files, but they all need to be together in one pull request.

**Comment by rjchow** [comment]
why do they need to be together?

**Comment by khunshwephyo** [comment]
> 
> why do they need to be together?
>

- TP needs to issue this year's part-time certificates in one batch, and then uploading them into MySkillsFuture portal, followed by sending notification emails with .opencert file attachments to students.  So all certificate templates need to be viewable in [https://opencerts.io](url).  
- Part-time templates are also very similar to one another.  I had to refactor common files many times and I wanted to avoid creating many pull requests for insignificantly small yet frequent changes.
- TP internal team also need to track which pull request corresponds with what new enhancements made to our internal system.  

Hence one pull request instead of spreading into multiple pull requests.  I hope this does not cause too much trouble to your team.

**Comment by rjchow** [comment]
74 files is way too many for us to review properly, please split it up
you can do the announcements after the PRs are all merged.

if you would like, we can merge the PRs in one batch when they have all been approved, as well

**Comment by rjchow** [comment]
why do they need to be together?

**Comment by khunshwephyo** [comment]
> 
> 74 files is way too many for us to review properly, please split it up
> you can do the announcements after the PRs are all merged.
> 
> if you would like, we can merge the PRs in one batch when they have all been approved, as well
>

I have reduced the files down to 23 files for this pull request.  I will submit subsequent pull requests in smaller chunks.

**Comment by rjchow** [comment]
thanks phyo appreciate it

**Comment by khunshwephyo** [comment]
Created 11 new pull requests, each containing 1 or 2 templates only.  Each pull request can be reviewed merged independently from one another.

File: src/components/CertificateTemplates/tlds/sg/edu/tp/common/certificateNameSection.js [hunk 1]
```diff
@@ -1,38 +1,43 @@
 import PropTypes from "prop-types";
 
 const CertificateNameSection = ({ certificate }) => {
-  const firstLine = "Diploma in";
-  const secondLine = certificate.name.replace(/diploma in/i, "").trim();
+  const splitTwoLines = value => {
+    const index = value.search(/ in /i);
```
Comment (rjchow) [inline src/components/CertificateTemplates/tlds/sg/edu/tp/common/certificateNameSection.js]: you can do 
```
> "diploma in astinrology in asd".split(" in ")
[ 'diploma', 'astinrology', 'asd' ]
```

File: src/components/CertificateTemplates/tlds/sg/edu/tp/ptdip/sample.opencert [hunk 2]
```diff
@@ -1,159 +1,194 @@
 {
```
Comment (rjchow) [inline src/components/CertificateTemplates/tlds/sg/edu/tp/ptdip/sample.opencert]: if you've already issued this type of certs I would recommend not making changes to the sample certs as it won't be reflective of the actual production certs - but if I remember correctly these certs haven't been issued?

Comment (khunshwephyo) [reply src/components/CertificateTemplates/tlds/sg/edu/tp/ptdip/sample.opencert]: The sample certs were changed to include new part-time transcript data.  You are right, these type of certs have not been issued yet.

Comment (khunshwephyo) [reply src/components/CertificateTemplates/tlds/sg/edu/tp/ptdip/sample.opencert]: The sample certs needed to change due to new data in transcript.  It is specific to part-time certs.  You are right, this type of certs haven't been issued yet.

File: src/components/CertificateTemplates/tlds/sg/edu/tp/common/certificateNameSection.js (hunk 1 above)
Comment (khunshwephyo) [reply src/components/CertificateTemplates/tlds/sg/edu/tp/common/certificateNameSection.js]: Thanks.  Simplified the method to use `.split()`

```JavaScript
  const splitTwoLines = value => {
    const lines = value.split(/ in (.+)/i);
    if (lines.length > 1) {
      return (
        <span>
          {lines[0]} in
          <br />
          {lines[1]}
        </span>
      );
    }
    return <span>{value}</span>;
  };
```

Comment (khunshwephyo) [reply src/components/CertificateTemplates/tlds/sg/edu/tp/common/certificateNameSection.js]: Also cherry-pick commit these changes to the rest of my pull requests.

**Comment by rjchow** [comment]
would you like me to merge this? it's approved

**Comment by khunshwephyo** [comment]
> 
> 
> would you like me to merge this? it's approved
>

Yes, please.  I would like to merge my pull requests as soon as approved.  Thanks!

**Comment by khunshwephyo** [comment]
I saw the progressive merging of my pull requests.  Thank you so much!

**Comment by rjchow** [comment]
all done :) do have a look over and make sure everything works as expected

**Comment by khunshwephyo** [comment]
> 
> all done :) do have a look over and make sure everything works as expected
>

Yup, everything works as expected.  :-)

**Pull request merged**